spec/rails_helper.rb

```diff
@@ -48,20 +47,69 @@
   Capybara::Selenium::Driver.new app, browser: :chrome, options: options
 end
 
-if EnvUtilities.load_boolean(name: 'HEADLESS', default: false)
-  # Run the feature specs in a full browser (note, this takes over your computer's focus)
-  Capybara.javascript_driver = :selenium_chrome_headless
+# The webdrivers gem uses selenium-webdriver.  Our docker approach needs selenium-webdriver
+# but gets upset if webdriver is loaded.  So in the Gemfile, we `require: false` both of
+# these and explicitly require them based on where we're running.  We also only register
+# the docker flavor of the driver if we are indeed running in docker.
+
+if in_docker?
+  require 'selenium-webdriver'
+
+  Capybara.register_driver :selenium_chrome_headless_in_docker do |app|
+      chrome_capabilities = ::Selenium::WebDriver::Remote::Capabilities.chrome(
+        'goog:chromeOptions' => { 'args': %w[no-sandbox headless disable-gpu] }
+      )
+
+      Capybara::Selenium::Driver.new(app,
+                                     browser: :remote,
+                                     url: ENV['HUB_URL'],
+                                     desired_capabilities: chrome_capabilities)
+  end
+
+  Capybara.javascript_driver = :selenium_chrome_headless_in_docker
+
+  # Normally the Capybara host is 'localhost', but within Docker it may not be.
+  capybara_host = IPSocket.getaddress(Socket.gethostname)
+  capybara_port = ENV.fetch("PORT") { DEV_URL_OPTIONS[:port] }
+
+  Capybara.asset_host = "http://#{capybara_host}:#{capybara_port}"
+  Capybara.app_host = "http://#{capybara_host}:#{capybara_port}"
+  Capybara.server_host = capybara_host
+  Capybara.server_port = capybara_port
 else
-  Capybara.javascript_driver = :selenium_chrome
-end
+  require 'webdrivers/chromedriver'
 
-Capybara.asset_host = 'http://localhost:2999'
+  if EnvUtilities.load_boolean(name: 'HEADLESS', default: false)
+    # Run the feature specs in a full browser (note, this takes over your computer's focus)
```

Comment:
what's the reason to not run local tests headless too?

Reply:
to actually see how things render, makes your chrome open and close a million times, but we can use the `screenshot!` to take pictures of every screen or we can use the debugger to stop in one screen and visually inspect it.